### PR TITLE
Add login errors

### LINKED ISSUE #601: Handle login errors
## Description
Handle error responses on failed logins. Display correct, localised error messages to the user. See Xamarin app to look at the possible errors.
## Dependencies
#597 
## Definition of done
Reducer sets the correct error message
UI displays the message
- [ ] The documentation is updated to reflect the changes done in this PR

### LINKED ISSUE #605: Implement account locking
## Description
Read the remaining attempts header on failed logins. When there’s one more try left, display the appropriate message. When the account is locked, display the appropriate message as well. All the details can be found in https://github.com/toggl/mobileapp/projects/109

## Definition of done
The remaining attempts header is read
The two messages are displayed accordingly
The info about remaining login attempts can be added in SetError, but if you have better ideas, go ahead

### PR DESCRIPTION
**Pull request by dshr**
## Info
This PR adds all the appropriate login errors, including the locked account ones. Also adds the sign up errors, which have no issue for some reason.

### Relationships
Closes #605
Closes #601

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Commit by commit will do it

### Review template
- [x] The code is thoroughly tested in well written and resilient manner
- [x] The code is performant and leads to an optimal user experience
- [x] The code is testable and does not cut corners for the sake of delivering
- [x] The code considers older platforms well and will work in older devices just like it will in new ones

## Merge permissions
Go for it!

### CODE REVIEW

**Comment by sofokli1** [comment]
#601 says 
```
- Reducer sets the correct error message
- UI displays the message
```
- I do not see any changes to reducers 
- The ui part should be implemented in PR #674 (but the PR mentions that it is not hooked to anything)

File: Modules/API/API/API.swift [hunk 1]
```diff
@@ -138,6 +138,25 @@ extension API: UserAPI {
...
         return urlSession.load(endpoint)
+            .catchError { error in
+                guard let error = error as? NetworkingError else { return Observable.error(APIError.loginGeneral) }
+                switch error {
+                case .noData:
+                    return Observable.error(APIError.offline)
+                case .unknown:
+                    return Observable.error(APIError.loginGeneral)
+                case let .wrongStatus(_, response):
+                    guard let remainingAttempts = response?.allHeaderFields["X-Remaining-Login-Attempts"] as? Int32
```
Comment (simonrozsival) [inline Modules/API/API/API.swift:149]: The header name could either be a constant or there could be an extension method to get the number of remaining login attempts from the response.

File: Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift [hunk 2]
```diff
@@ -124,16 +124,66 @@ public class LoginViewController: UIViewController, Storyboarded {
...
+            passwordErrorText = error.description
+            shouldShowPasswordError = true
+        }
+
+        emailErrorLabel.text = emailErrorText
+        emailErrorLabel.isHidden = !shouldShowEmailError
+        emailTextField.isValid = !shouldShowEmailError
+        passwordErrorLabel.text = passwordErrorText
+        passwordErrorLabel.isHidden = !shouldShowPasswordError
+        emailTextField.isValid = !shouldShowEmailError
```
Comment (sofokli1) [inline Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift:165]: this line is the same with 162

File: Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift [hunk 3]
```diff
@@ -124,16 +124,60 @@ public class LoginViewController: UIViewController, Storyboarded {
...
         let signUpButtonTitle = questionAndCTA(Strings.Login.signUpQuestion, Strings.Login.signUpCTA)
         signUpButton.setAttributedTitle(signUpButtonTitle, for: .normal)
     }
-    
+
+    private func handleUser(user: Loadable<User>) {
+        var emailErrorText = ""
+        var passwordErrorText = ""
+        var shouldShowEmailError = false
+        var shouldShowPasswordError = false
```
Comment (sofokli1) [inline Modules/Onboarding/Onboarding/EmailLogin/Views/LoginViewController.swift:140]: all these `var`s make this more complicated for not much gain. you could just set the `emailErrorLabel` and `passwordErrorLabel` to hidden at the top of the function, and do what is happening in lines 154-159 in the `if case`s directly without the need for extra `var`s

File: Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift [hunk 4]
```diff
@@ -113,9 +118,58 @@ public class SignupViewController: UIViewController, Storyboarded {
...
-        passwordTextField?.resignFirstResponder()
+        _ = emailTextField?.resignFirstResponder()
+        _ = passwordTextField?.resignFirstResponder()
     }
+    
+    private func handleUser(user: Loadable<User>) {
+        var emailErrorText = ""
+        var passwordErrorText = ""
+        var shouldShowEmailError = false
+        var shouldShowPasswordError = false
```
Comment (sofokli1) [inline Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift:134]: same point as in the login one

File: Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift [hunk 5]
```diff
@@ -113,9 +118,58 @@ public class SignupViewController: UIViewController, Storyboarded {
...
+            passwordErrorText = error.description
+            shouldShowPasswordError = true
+        }
+
+        emailErrorLabel.text = emailErrorText
+        emailErrorLabel.isHidden = !shouldShowEmailError
+        emailTextField.isValid = !shouldShowEmailError
+        passwordErrorLabel.text = passwordErrorText
+        passwordErrorLabel.isHidden = !shouldShowPasswordError
+        emailTextField.isValid = !shouldShowEmailError
```
Comment (sofokli1) [inline Modules/Onboarding/Onboarding/EmailSignup/Views/SignupViewController.swift:159]: same point as in the login one

**Comment by dshr** [comment]
It's a binding issue, I copy-pasted things without double-checking

**Comment by dshr** [comment]
That's a good catch right there — I've disabled the submit button until the password is strong, so tapping it just hides the keyboard, which makes things look broken. I'll figure out a way around it.

**Comment by dshr** [comment]
@sofokli1 have another look, I've switched the UX around a bit, this should work better.

**Pull request merged**